File: 3_netcat/netcat.py
#!/usr/bin/python

# 使用 IO 复用实现 netcat

import os
import select
import socket
import sys

def relay(sock):
    poll = select.poll()
    poll.register(sock, select.POLLIN)      # 为 sock 注册读事件
    poll.register(sys.stdin, select.POLLIN) # 为 stdin 注册读事件

    done = False
    while not done:
        events = poll.poll(1000)  # 10 seconds, 等待事件
        for fileno, event in events:
            if event & select.POLLIN:
                if fileno == sock.fileno():  # socket 可读
                    data = sock.recv(8192)
                    if data:
                        sys.stdout.write(data.decode('utf-8'))  # call c 中的 fwrite()
                    else:                       # 读到 0 就退出，不需要像 netcat.cc 那样还要通知另外的线程
                        done = True             # 数据读完才能关闭链接，否则会使 TCP 发送 RST 分节，导致数据接收不完整
                else:
                    assert fileno == sys.stdin.fileno()
                    data = os.read(fileno, 8192)
                    if data:
                        sock.sendall(data)
                    else:
                        sock.shutdown(socket.SHUT_WR)
                        poll.unregister(sys.stdin)
                        # 不直接 done = True，是需要等数据发送完，否则数据发送不完整（as nc_sender.cc)


def main(argv):
    if len(argv) < 3:
        binary = argv[0]
        print("Usage:\n  %s -l port\n  %s host port" % (argv[0], argv[0]))
        return
    port = int(argv[2])
    if argv[1] == "-l":
        # server
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind('', port)
        server_socket.listen(5)
        (client_socket, client_address) = server_socket.accept()
        server_socket.close()
        relay(client_socket)
    else:
        # client
        sock = socket.create_connection((argv[1], port))
        relay(sock)


if __name__ == "__main__":
    main(sys.argv)

File: 3_netcat/netcat.cc
#include <iostream>
#include <thread>
#include <vector>

#include "sylar/socket.h"
#include "util/util.h"

// netcat implementation base on thread-per-connection

void run(sylar::Socket::ptr sock) {
    // caution: a bad example for closing connection
    std::thread thr([&sock](){  // 网络线程 start: 网络线程不断地 socket 读，并将读到的数据发送到 stdout
        char buf[8192];
        int nr = 0;
        while ( (nr = sock->recv(buf, sizeof(buf))) > 0) {
            int nw = Util::WriteFixedBytes(STDOUT_FILENO, buf, nr);
            if (nw < nr) {
                break;
            }
        }
        ::exit(0);  // should somehow notify main thread instead.
    }); // 网络线程 end

    // 主线程：不断地从 stdin 读 8K 缓冲区，并将读到的数据发送到 socket
    char buf[8192];
    int nr = 0;
    while ( (nr = ::read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
        int nw = Util::WriteFixedBytes(sock, buf, nr);
        if (nw < nr) {
            break;
        }
    }
    sock->shutdownWrite();
    thr.join();
}

void run_grace(sylar::Socket::ptr sock) {
    // use select() to exit gracefully

    int pipefds[2];
    int rt = pipe(pipefds);
    assert(!rt);

    std::thread thr([&sock, &pipefds](){  // 网络线程 start: 网络线程不断地 socket 读，并将读到的数据发送到 stdout
        char buf[8192];
        int nr = 0;
        while ( (nr = sock->recv(buf, sizeof(buf))) > 0) {
            int nw = Util::WriteFixedBytes(STDOUT_FILENO, buf, nr);
            if (nw < nr) {
                break;
            }
        }
        int rt = ::write(pipefds[1], "t", 1);
        assert(rt == 1);
        sock->shutdownRead();
    }); // 网络线程 end

    // 主线程：不断地从 stdin 读 8K 缓冲区，并将读到的数据发送到 socket
    char buf[8192];
    int nr = 0;
    fd_set readset;
    int result = 0;
    do {
        FD_ZERO(&readset);
        FD_SET(STDIN_FILENO, &readset);
        FD_SET(pipefds[0], &readset);

        result = select(std::max(STDIN_FILENO, pipefds[0]) + 1, &readset, nullptr, nullptr, nullptr);
        if (result > 0) {
            if (FD_ISSET(STDIN_FILENO, &readset)) {
                nr = ::read(STDIN_FILENO, buf, sizeof(buf));
                if (nr > 0) {
                    int nw = Util::WriteFixedBytes(sock, buf, nr);
                    if (nw < nr) {
                        break;
                    }
                } else {  // 读 stdin 结束 
                    break;
                }
            }

            if (FD_ISSET(pipefds[0], &readset)) {
                // 套接字已关闭，sock->recv() 得到 0
                nr = sock->recv(buf, sizeof(buf));
                if (nr <= 0) {
                    break;  // 套接字关闭
                }
            }
        }
    } while (result != -1);
    
    sock->shutdownWrite();
        
    thr.join();
    close(pipefds[0]);
    close(pipefds[1]);
}

int main(int argc, char const *argv[]) {
    if (argc < 3) {
        std::cout << "Usage:" << std::endl 
                << "  " << argv[0] << " hostname port" << std::endl
                << "  " << argv[0] << " -l port" << std::endl; 
        return 0;
    }

    int port = atoi(argv[2]);
    if (strcmp(argv[1], "-l") == 0) {       // as server
        std::string host = "localhost:" + std::to_string(port);
        auto addr = sylar::Address::LookupAny(host);
        if (addr) {
            auto listen_sock = sylar::Socket::CreateTCPSocket();
            listen_sock->bind(addr);
            listen_sock->listen();
            auto sock = listen_sock->accept();
            if (sock) {
                listen_sock.reset();    // stop listening
                // run(sock);              // unique_ptr more suitable
                run_grace(sock);
            } else {
                perror("accept");
            }
        } else {
            std::cout << "Unable reslove " << host << std::endl;
            return 0;
        }
    } else {        // as client
        std::string hostname = std::string(argv[1]) + ":" + std::to_string(port);
        auto addr = sylar::Address::LookupAny(hostname);
        if (addr) {
            auto client_sock = sylar::Socket::CreateTCPSocket();
            client_sock->connect(addr);
            if (client_sock) {
                // run(client_sock);
                run_grace(client_sock);
            } else {
                std::cout << "Unable to connect " << *addr << std::endl;
            }
        } else {
            std::cout << "Unable reslove " << hostname << std::endl;
        }
    }
    return 0;
}
